docs/proposals/user-namespace.md

```diff
@@ -0,0 +1,143 @@
+for monitoring.  Assuming that the docker daemon is started with
+`--userns-remap=default`,
+here is when the host user namespace should be set:
+
+1.  When setting any other namespace flag to `host`
+    * Not strictly required for `net=host` if not binding to low ports
+     but if the system provides defaulting it covers the 80% case
+2.  When using capabilities that aren't tied to any namespace
+    * MKNOD
+    * SYS_MODULE
+    * SYS_TIME
+    * others?  (@mrunalp is looking at source to determine this)
+3.  When explicitly granted permissions to use a host volume and
+escape the private user namespace
+
+
+### Concrete Examples
+
+| Use Case  | Support |
+|---|---|
+| Load a kernel module  | --userns=host --cap-add SYS_ADMIN  |
+| Create a device using mknod  |  --userns=host --cap-add MKNOD |
+| Access host path as root user  | --userns=host --cap-add SYS_ADMIN  |
+| Bind to a low port on a host interface  | --userns=host, --net=host*  |
+| Monitor other processes on host looking at /proc  | --userns=host, --pid=host*  |
+| Bind to a high port on a host interface | --net=host, --userns=container_private^  |
+| Share net,pid namespaces between containers  | --net,--pid=container:id implicitly shares userns as well  |
+
+We can make a couple of generalizations from the table:
+
+1. Certain capabilities that aren't tied to any namespace such as
+CAP_MKNOD, CAP_SYS_MODULE or CAP_SYS_TIME are requested, then they
+imply userns=host.
+2. Setting any other namespace to host could imply userns=host.
+
+[*] = requires docker/docker#25771
+[^] = not supported in docker today but could be done using runc
+
+
+
+## Options
+
+1.  Provide a `HostUser` flag and require manual setting.
+
+    Pros
+    *  Quick and easy
+    *  Allows full flexibility
+
+    Cons
+    *  Requires linux specific knowledge
+
+2.  Provide a `HostUser` flag and provide auto setting.
+
+    Pros
+    *  System does the right thing when it can
+
+    Cons
+    *  Requires extra flags/config - this defaulting should only occur
+    when remapping is configured on the Docker daemon.
+    *  Does not support edge cases like using user namespaces but binding
+    to high ports on the host.
+    *  The user may unexpectedly run a container with the host user namespace.
+
+## Proposal
+
+For 1.4 we can add the `HostUser` flag to provide immediate support
+for anyone who has a use case that is not currently supported and
+enable users on Docker 1.10+ to test and enable `userns=remap`.  We can
+then iterate on smart defaulting in future releases to ensure that the
+system will do the right thing when remapping is known to be configured.
+
+
+## Additional Concerns
+
+When any new field is added to the `SecurityContext` it should also be
+added to the `PodSecurityPolicy` so that it may be controlled by an
+administrator.
+
+In this case the field is a boolean and should be controlled by an option
+that allows it to be set or not.  PSP should not provide a default.
```

Comment:
I've said my peace.  I STILL don't feel like I have my head wrapped around
it, so I fully expect this to suck for users, too.  It's pretty niche, I
hope

On Sun, Aug 21, 2016 at 10:37 AM, Clayton Coleman <notifications@github.com>
wrote:

> In docs/proposals/user-namespace.md
> <https://github.com/kubernetes/kubernetes/pull/30684#discussion_r75599648>
> :
>
> > +
> > +For 1.4 we can add the `HostUser` flag to provide immediate support
> > +for anyone who has a use case that is not currently supported and
> > +enable users on Docker 1.10+ to test and enable `userns=remap`.  We can
> > +then iterate on smart defaulting in future releases to ensure that the
> > +system will do the right thing when remapping is known to be configured.
> > +
> > +
> > +## Additional Concerns
> > +
> > +When any new field is added to the `SecurityContext` it should also be
> > +added to the `PodSecurityPolicy` so that it may be controlled by an
> > +administrator.
> > +
> > +In this case the field is a boolean and should be controlled by an option
> > +that allows it to be set or not.  PSP should not provide a default.
>
> +For 1.4 we can add the `HostUser` flag to provide immediate support +for
> anyone who has a use case that is not currently supported and +enable users
> on Docker 1.10+ to test and enable `userns=remap`. We can +then iterate on
> smart defaulting in future releases to ensure that the +system will do the
> right thing when remapping is known to be configured. + + +## Additional
> Concerns + +When any new field is added to the `SecurityContext` it should
> also be +added to the `PodSecurityPolicy` so that it may be controlled by
> an +administrator. + +In this case the field is a boolean and should be
> controlled by an option +that allows it to be set or not. PSP should not
> provide a default.
> PSP typically allows all operations right? So disabling userns would be
> the default behavior right? It does in Kube, but I think that might be
> wrong in general. OpenShift defaults to safe by default by putting all
> restrictions on unless you escalate - I think we want that for Kube
> eventually. We could argue that defaulting on makes sense in the immediate
> timeframe for Kube, but since you have to turn userns on also, it seems
> reasonable to default to off in the API and let the Kubelet/runtime provide
> that explicit behavior.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30684/files/e535458bb0d3cf06fd7b585c1f583e26f2d684c7#r75599648>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVGSC2TlNgZtgxXOrY94YLQ562iTEks5qiIy-gaJpZM4Jlddf>
> .
>